drivers/mdio/mdio_gpio.c

```diff
@@ -0,0 +1,195 @@
+{
+
+	gpio_pin_configure_dt(&dev_cfg->mdio_gpio, (dir) ? GPIO_OUTPUT_ACTIVE : GPIO_INPUT);
+	if ((dir) == 0) {
+		mdio_gpio_clock_the_bit(dev_cfg);
+	}
+}
+
+static ALWAYS_INLINE void mdio_gpio_read(dev_cfg_ptr dev_cfg, uint16_t *pdata)
+{
+	uint16_t data = 0;
+
+	for (uint16_t i = 0; i < 16; i++) {
+		data <<= 1;
+		mdio_gpio_clock_the_bit(dev_cfg);
+		if (gpio_pin_get_dt(&dev_cfg->mdio_gpio) == 1) {
+			data |= 1;
+		}
+	}
+	*pdata = data;
+}
+
+static ALWAYS_INLINE void mdio_gpio_write(dev_cfg_ptr dev_cfg, uint32_t data, uint8_t len)
+{
+	uint32_t v_data = (data);
+	uint32_t v_len = (len);
+
+	v_data <<= (32 - v_len);
+	for (; v_len > 0; v_len--) {
+		gpio_pin_set_dt(&dev_cfg->mdio_gpio, (v_data & MDIO_GPIO_MSB) ? 1 : 0);
+		mdio_gpio_clock_the_bit(dev_cfg);
+		v_data <<= 1;
+	}
+}
+
+static int mdio_gpio_transfer(const struct device *dev, uint8_t prtad, uint8_t devad, uint8_t rw,
+			      uint16_t data_in, uint16_t *data_out)
+{
+	const struct mdio_gpio_config *const dev_cfg = dev->config;
+	struct mdio_gpio_data *const dev_data = dev->data;
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+
+	/* DIR: output */
+	mdio_gpio_dir(dev_cfg, MDIO_GPIO_WRITE_OP);
+	/* PRE32: 32 bits '1' for sync*/
+	mdio_gpio_write(dev_cfg, 0xFFFFFFFF, 32);
+	/* ST: 2 bits start of frame */
+	mdio_gpio_write(dev_cfg, 0x1, 2);
+	/* OP: 2 bits opcode, read '10' or write '01' */
+	mdio_gpio_write(dev_cfg, rw ? 0x1 : 0x2, 2);
+	/* PA5: 5 bits PHY address */
+	mdio_gpio_write(dev_cfg, prtad, 5);
+	/* RA5: 5 bits register address */
+	mdio_gpio_write(dev_cfg, devad, 5);
+
+	if (rw) { /* Write data */
+		/* TA: 2 bits turn-around */
+		mdio_gpio_write(dev_cfg, 0x2, 2);
+		mdio_gpio_write(dev_cfg, data_in, 16);
+	} else { /* Read data */
+		/* Release the MDIO line */
+		mdio_gpio_dir(dev_cfg, MDIO_GPIO_READ_OP);
+		mdio_gpio_read(dev_cfg, data_out);
+	}
+
+	/* DIR: input. Tristate MDIO line */
+	mdio_gpio_dir(dev_cfg, MDIO_GPIO_READ_OP);
+
+	k_sem_give(&dev_data->sem);
+
+	return 0;
+}
+
+static int mdio_gpio_read_mmi(const struct device *dev, uint8_t prtad, uint8_t devad,
+			      uint16_t *data)
+{
+	return mdio_gpio_transfer(dev, prtad, devad, MDIO_GPIO_READ_OP, 0, data);
+}
+
+static int mdio_gpio_write_mmi(const struct device *dev, uint8_t prtad, uint8_t devad,
+			       uint16_t data)
+{
+	return mdio_gpio_transfer(dev, prtad, devad, MDIO_GPIO_WRITE_OP, data, NULL);
+}
+
+static int mdio_gpio_initialize(const struct device *dev)
+{
+	const struct mdio_gpio_config *const dev_cfg = dev->config;
+	struct mdio_gpio_data *const dev_data = dev->data;
+	int rc;
+
+	k_sem_init(&dev_data->sem, 1, 1);
+
+	if (!device_is_ready(dev_cfg->mdc_gpio.port)) {
+		LOG_ERR("GPIO port for MDC pin is not ready");
+		return -ENODEV;
+	}
+	rc = gpio_pin_configure_dt(&dev_cfg->mdc_gpio, GPIO_OUTPUT_INACTIVE);
+	if (rc < 0) {
+		LOG_ERR("Couldn't configure MDC pin; (%d)", rc);
+		return rc;
+	}
+
+	if (!device_is_ready(dev_cfg->mdio_gpio.port)) {
```

Comment:
This check should better to place before previous `before gpio_pin_configure_dt`.
It can reduce modification possibility of gpio status even if errors.

Reply:
Fixed